### PR TITLE
cancel: add a controller that adds cancel buttons

### PR DESCRIPTION
**Pull request by nicks**
Hello @landism, @lucieleblanc,

Please review the following commits I made in branch nicks/ch12404:

487f997c82d8a65aaeacb8750f50e323a38ae462 (2021-07-20 11:34:32 -0400)
cancel: add a controller that adds cancel buttons

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
This is an example of the kind of thing I'd like to be installable as a global extension as in https://github.com/tilt-dev/tilt.specs/pull/30

I tagged @lucieleblanc on this since she's working on a similar controller that dynamically adds buttons, so there might be some good patterns to establish on how those buttons store state, and how they set up actions

File: cancel/reconcile_cancel_btn.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    delete_children
+fi
+
+# Ignore commands that don't belong to a resource.
+resource=$(echo "$cmd" | jq -r '.metadata.annotations["tilt.dev/resource"]')
+if [[ "$resource" == "null" ]]; then
+    exit 0
+fi
+
+# Ignore cmds that don't to the Tiltfile or a local_resource.
```
Comment (ellenkorbes) [inline cancel/reconcile_cancel_btn.sh:35]:
```suggestion
# Ignore cmds that don't belong to the Tiltfile or a local_resource.
```
?

**Comment by ellenkorbes** [comment]
Gonna leave reviewing the code to others but 10/10 on concept and structure.

**Review by landism — APPROVED** [review]
LGTM

Some nice-to-haves, but not blockers:
1. a test.sh that runs shellcheck, which we probably want in the ci image for other extensions and also should encourage for people using this as an example
2. it also might not be that expensive to write actual tests for the reconciliation? they'd serve as nice documentation and guardrails for people iterating on their own extensions using this as an example, but that's also a fair amount of extra scope for this and maybe not worth it yet.

File: cancel/reconcile_cancel_btn.sh (hunk 1 above)
Comment (nicks) [reply cancel/reconcile_cancel_btn.sh:35]: yep haha

File: cancel/reconcile_cancel_btn.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+owner_kind=$(echo "$cmd" | jq -r '.metadata.labels["tilt.dev/owner-kind"]')
+echo "kind 1 $owner_kind"
+if [[ "$owner_kind" == "null" ]]; then
+    # gahhhhh
+    owner_kind=$(echo "$cmd" | jq -r '.metadata.annotations["tilt.dev/owner-kind"]')
+    echo "kind 2 $owner_kind"
```
[the remark this answers is not in the document]
Comment (nicks) [reply cancel/reconcile_cancel_btn.sh:41]: gahhhh yes

File: cancel/reconcile_cancel_btn.sh [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+fi
+if [[ "$owner_kind" != "Tiltfile" && "$owner_kind" != "CmdServer" ]]; then
+    exit 0
+fi
+
+# If the command isn't running or doesn't exist, delete any existing button.
```
[the remark this answers is not in the document]
Comment (nicks) [reply cancel/reconcile_cancel_btn.sh:47]: yes! dogfooding is great! I'll add a story for that

**Comment by nicks** [comment]
ya, i'm going to do tests in a follow-up. 

reconciliation loops make it really easy to write tests!

but bash makes it harder to write tests :joy:

File: cancel/reconcile_cancel_btn.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicks) [reply cancel/reconcile_cancel_btn.sh:45]: https://app.clubhouse.io/windmill/story/12408/add-disable-field-to-button-spec

**Pull request merged**